python/lsst/meas/astrom/pessimistic_pattern_matcher_b_3D.py

```diff
@@ -0,0 +1,1111 @@
+        if sin_shift > 0:
+            rot_axis = np.cross(src_center, ref_center)
+            rot_axis /= sin_shift
+
+            rot_cross_matrix = np.array(
+                [[0., -rot_axis[2], rot_axis[1]],
+                 [rot_axis[2], 0., -rot_axis[0]],
+                 [-rot_axis[1], rot_axis[0], 0.]], dtype=np.float64)
+            shift_matrix = (cos_shift*np.identity(3) +
+                            sin_shift*rot_cross_matrix +
+                            (1. - cos_shift)*np.outer(rot_axis, rot_axis))
+        else:
+            shift_matrix = np.identity(3)
+
+        # Now that we have our shift we apply it to the src delta vector
+        # and check the rotation.
+        rot_src_delta = np.dot(shift_matrix, src_delta)
+        cos_rot_sq = (np.dot(rot_src_delta, ref_delta)**2 /
+                      (np.dot(rot_src_delta, rot_src_delta) *
+                       np.dot(ref_delta, ref_delta)))
+        # If the rotation isn't in tolerance return None.
+        if cos_rot_sq < max_cos_rot_sq:
+            return None
+        return pipeBase.Struct(
+            cos_rot_sq=cos_rot_sq,
+            shift_matrix=shift_matrix,)
+
+    def _create_pattern_spokes(self, src_ctr, src_delta_array, src_dist_array,
+                               ref_ctr, ref_ctr_id, ref_delta, ref_dist,
+                               ref_delta_array, ref_dist_array,
+                               ref_id_array, max_dist_rad, n_match):
+        """ Create the indiviual spokes that make up the pattern now that the
+        shift and rotation are within tolerance. If we can't create a valid
+        pattern we exit early.
+
+        Parameters
+        ----------
+        src_ctr : float array
+            3 vector of the source pinwheel center
+        src_delta_array : float array
+            Array of 3 vector deltas between the source center and the pairs
+            that make up the remaining spokes of the pinwheel
+        src_dist_array : float array
+            Array of the distances of each src_delta in the pinwheel
+        ref_ctr : float array
+            3 vector of the candidate refenerce center
+        ref_ctr_id : int
+            id lookup of the ref_ctr into the master reference array
+        ref_delta : float array
+            3 vector of the first candidate pair of the pinwheel. This is
+            the candidate pair that was matched in the
+            main _construct_pattern_and_shift_rot_matrix loop
+        ref_dist : float
+            Radian distance of the first candidate reference pair
+        ref_delta_array : float array
+            Array of 3 vector deltas that are have the current candidate
+            reference center as part of the pair
+        ref_dist_array : float array
+            Array of vector distances for each of the reference pairs
+        ref_id_array : int array
+            Array of pair id lookups into the master reference array. All
+            pairs contain the reference center id.
```

Comment:
I suppose I could also check to see if the candidate id is less than that of the id of the center, as that would tell me the direction of the delta vector. This reduces the size of the pair_id_array by half so seems like an upgrade. Change made and doc altered.